test/framework/admission-webhooks.go

```diff
@@ -23,40 +23,66 @@ import (
 	"k8s.io/apimachinery/pkg/util/yaml"
 )
 
-func (f *Framework) createMutatingHook(ctx context.Context, certBytes []byte, namespace, yamlPath string) (FinalizerFn, error) {
-	h, err := parseMutatingHookYaml(yamlPath)
+func (f *Framework) createOrUpdateMutatingHook(ctx context.Context, certBytes []byte, namespace, source string) (FinalizerFn, error) {
+	hook, err := parseMutatingHookYaml(source)
 	if err != nil {
 		return nil, errors.Wrap(err, "Failed parsing mutating webhook")
 	}
 
-	h.Webhooks[0].ClientConfig.Service.Namespace = namespace
-	h.Webhooks[0].ClientConfig.CABundle = certBytes
-
-	_, err = f.KubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Create(ctx, h, metav1.CreateOptions{})
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to create mutating webhook %s", h.Name)
+	hook.Webhooks[0].ClientConfig.Service.Namespace = namespace
+	hook.Webhooks[0].ClientConfig.CABundle = certBytes
+
+	h, err := f.KubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, hook.Name, metav1.GetOptions{})
+	if err == nil {
+		// must set this field from existing MutatingWebhookConfiguration to prevent update fail
+		hook.ObjectMeta.ResourceVersion = h.ObjectMeta.ResourceVersion
+
+		// MutatingWebhookConfiguration already exists -> Update
+		_, err = f.KubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Update(ctx, hook, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to update mutating webhook %s", hook.Name)
+		}
+	} else {
+		// MutatingWebhookConfiguration doesn't exists -> Create
+		_, err = f.KubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Create(ctx, hook, metav1.CreateOptions{})
```

Comment:
@PhilipGough  I agree, I will do it. But currently I have some question regarding the error message part. I saw two different pattern when we want to return the error. [One](https://github.com/prometheus-operator/prometheus-operator/blob/main/test/framework/service_account.go#L34) will return the error directly, and [another ](https://github.com/prometheus-operator/prometheus-operator/blob/main/test/framework/deployment.go#L56) will wrap the error with additional info before return it. May I know I should follow which pattern?

Reply:
Personally, I think since we are using third party library, and we do not know the actual error message is clear enough or not, so I think is better to wrap the error with some additional info?

Reply:
@PhilipGough understand, then I remained the way to return error. It is done as well, please have a look :)